packages/react/src/ReactAct.js

```diff
@@ -91,21 +91,30 @@ export function act<T>(callback: () => T | Thenable<T>): Thenable<T> {
 
       if (__DEV__) {
         if (!didWarnNoAwaitAct && typeof Promise !== 'undefined') {
-          // eslint-disable-next-line no-undef
-          Promise.resolve()
-            .then(() => {})
-            .then(() => {
-              if (!wasAwaited) {
-                didWarnNoAwaitAct = true;
-                console.error(
-                  'You called act(async () => ...) without await. ' +
-                    'This could lead to unexpected testing behaviour, ' +
-                    'interleaving multiple act calls and mixing their ' +
-                    'scopes. ' +
-                    'You should - await act(async () => ...);',
-                );
-              }
-            });
+          const awaitDetector = () => {
+            if (!wasAwaited) {
+              didWarnNoAwaitAct = true;
+              console.error(
+                'You called act(async () => ...) without await. ' +
+                  'This could lead to unexpected testing behaviour, ' +
+                  'interleaving multiple act calls and mixing their ' +
+                  'scopes. ' +
+                  'You should - await act(async () => ...);',
+              );
+            }
+          };
+          if (result instanceof Promise && 'finally' in result) {
+            // Using finally here is more compatible with native promises
+            // and promise libraries like core.js
+            result
+              .finally(() => {})
+              .catch(() => {})
+              .then(awaitDetector);
+          } else {
+            Promise.resolve()
+              .then(() => {})
+              .then(awaitDetector);
+          }
```

Comment:
```suggestion
          Promise.resolve(result)
            .then(() => {}, () => {})
            .then(awaitDetector);
```

Reply:
Awesome, thanks this is much cleaner and it fixes the original issue all of my local tests (existing RN repo, fresh RN repo, React Web).

TY!